### DISCUSSION TITLE
RPi Pico - use DMA to pass data to PIO repeatedly

### DISCUSSION THREAD

**Discussion by SteKme** [micropython/micropython#11081] [Using MicroPython]
Hi,

I wonder if there is a way to configure DMA (using utypes..) in a way that would allow me to pass same set of bytes to the PIO OSR register repeatedly - e.g. as in case of HDMI, VGA out, imagine there is a memory block (framebuffer) that needs to get sent to the PIO again and again.. - without the CPU's interception, any need to reset anything when the frame gets sent, simply start over and send bytes again..
-- in my case, it is just a couple of bytes, but with uPy, I'd really prefer not to check DMA current write registers repeatedly to ensure I reset them to the start of the memory block when the block is all sent to the PIO (
-- yes, I can allocate a rather large block that will have the same sequence copied many times to lower number of required DMA resets in time, but anyway.. 
-- and the problem is - to reset registers at the correct moment, which is almost impossible to achieve given the update speed of PIO vs low speed of uPy - even if I tested in a loop, I might be unable to reset the DMA fast enough to avoid timing glitches..
-- therefore - would there be some way to configure DMA to pass the same block of bytes to PIO repeatedly?

Thank You!
Stepan

**Comment by robert-hh** [comment]
You can configure DMA to repeat transmitting the same data to PIO over and over. There are a few PR ofr adding a DMA class to the RP2 port. A while ago I had made a few examples using DMA with PIO. https://github.com/robert-hh/RP2040-Examples/tree/master/rp2_util. These transmit once, but can easily extended to repeat. The driver uses viper code, which is more natural to deal with the hardware directly.

**Reply by rkompass** [reply]
I was fond of Roberts DMA and extended it a bit. Now it recognizes types of arguments and configures itself accordingly, if you do not enforce your presets. Also there is a way to repeatedly transfer memory blocks, like with sound (another instance of the DMA reconfigures the 'worker DMA'). No interrupts yet. Actually I need help to integrate interrupts as I'm not acquainted enough with MPs internals. ?@robert-hh? (you are already very busy I know).
It's [here](https://github.com/rkompass/RPi2040_mPy). For an example of reconfiguration (sound) see rp2_dma_test5.py.

A new example that outputs bytes:
This will output a binary array on Pin(0) .. Pin(7) with RP2_Dma through the PIO.
The example is slowed down to 1Hz, so you may observe your LEDs :-)
```
from array import array
from machine import Pin
from time import sleep_ms
from rp2 import PIO, StateMachine, asm_pio
from rp2_dma import RP2_Dma

#  PIO program that pulls from the FIFO, outputs that as a byte, then waits another 1998 instructions.
@asm_pio(out_init=(PIO.OUT_LOW,)*8, out_shiftdir=PIO.SHIFT_RIGHT, pull_thresh=8)
def pio_out8():
    pull(block)
    out(pins, 8)
    # Cycles: 1 + 1 + 1 + 13 + 32 * (61 + 1) = 2000
    set(x, 31)                  [12]
    label("delay_high")
    nop()                       [29]
    nop()                       [30]
    jmp(x_dec, "delay_high")

#  With freq=2000 this will output a byte every second.
sm = rp2.StateMachine(0, pio_out8, freq=2000, out_base=Pin(0))
sm.active(1)

arr = array('B', (i for i in range(256)))
dma = RP2_Dma(arr, sm) # RP2_Dma inspects and uses pull_thresh=8 in @asm_pio to set dsize=0, otherwise we would get a
                       #   ValueError: Conflicting source/destination data sizes: specify dsize explicitly
                       #   and we had to provide dsize=0 as argument here.
dma.start()
while dma.busy() > 0:
    print(dma.transfer_count())
    sleep_ms(1000)
```

**Reply by SteKme** [reply]
thank You @rkompass , I will certainly read through all this! 
Yea, as You write, repeated sound could be another example of this kind, very true, e.g. endless playback of a sine wave or whatnot.

I think I don't need interrupts - I just need the DMA to refill the TX FIFO of the PIO as the statemachine pulls 32bit words out of it to prevent underflow - pulling from the TX fifo is HW linked to the DMA and allows for this automated refill of the fifo I have read. 
(-- if I truly needed, I might trigger IRQ from the PIO asm...)

Cheers,
Stepan

**Reply by robert-hh** [reply]
Thank you, @rkompass , to pick up the baton. Your code looks good.

**Reply by RobertAtQuestacon** [reply]
I've been trying to use the example provided by @rkompass as the basis for a multichannel driver for APA102 LED strips. 

I couldn't get the example to work until I put wrap_target() and wrap() at the start and end of the state machine code. Should it have worked without?

Even then it wouldn't work most of the time. It would spit out a constant transfer count between 1 and 256.

I placed a print statement within the rp2_dma.py to check the calculated transfer size, then I requested the transfer count before and after starting the dma. This is an example of what I'd get:
...
tr_cnt 256
Prestart: 2
PostStart 2
2
2
2
... repeating

I found that by inserting sm.put(20) before starting the dma, I could nearly guarantee the dma would continue. I have also tried moving the sm.active(1) to after the dma.start() with some success (not not always).

```
Here is my current code:
#from https://github.com/orgs/micropython/discussions/11081
from array import array
from machine import Pin
from time import sleep_ms
from rp2 import PIO, StateMachine, asm_pio
from rp2_dma import RP2_Dma

#  PIO program that pulls from the FIFO, outputs that as a byte, then waits another 1998 instructions.
@asm_pio(out_init=(PIO.OUT_LOW,)*8, out_shiftdir=PIO.SHIFT_RIGHT, pull_thresh=8)
def pio_out8():
    wrap_target()
    pull(block)
    out(pins, 8)
    # Cycles: 1 + 1 + 1 + 13 + 32 * (61 + 1) = 2000
    set(x, 31)                  [12]
    label("delay_high")
    nop()                       [29]
    nop()                       [30]
    jmp(x_dec, "delay_high")
    wrap()

#  With freq=2000 this will output a byte every second.
sm = rp2.StateMachine(0, pio_out8, freq=2000, out_base=Pin(0))
sm.active(1)

arr = array('B', (i for i in range(256)))
sm.put(20)

dma = RP2_Dma(arr, sm) # RP2_Dma inspects and uses pull_thresh=8 in @asm_pio to set dsize=0, otherwise we would get a
                       #   ValueError: Conflicting source/destination data sizes: specify dsize explicitly
                       #   and we had to provide dsize=0 as argument here.
dma.start()
while dma.busy() > 0:
    print(dma.transfer_count())
    sleep_ms(1000)

```

By the by, I've also found that I can remove the pull(block) statement and use autopull instead. It makes no difference the the good or bad behaviour.

Can someone explain why the pio code (sometimes) needs to be kick started with a put command? Is the FIFO somehow clogged? Does the pio need to be stopped before another pio is closed (ie. if I press the Cntrl-C and then load a new program)?

Thanks in advance for any suggestions, and thankyou @rkompass for getting me started.

**Reply by rkompass** [reply]
I used Thonny to try out my examples and one of the conditions implied by this is that when running code in a window Thonny first does a soft reset.
So if you do a soft reset before running your example code, are the issues gone?

Alternatively you might before running your code do a `dma.deinit()` or `dma.restart()` together with
`sm.restart(); sm.active(1)`.

As far as I understand it, having the PIO execute a `pull(block)` without giving it data is like a wait instruction and stalls its program execution. You doing a `sm.put(20)` gives it data and so it can run another loop.
You may test this by trying out many such `sm.put(b)` instructions with different values of b and immediately afterwards the byte b should appear at the pins.
But also starting the DMA again should lead to a continuation of the PIO program as it tries to feed the PIO the data as fast as the PIO may digest it, using the internal data request mechanism. This way the PIO sets the speed of operation in my example.

The `wrap_target()` `wrap()` should be done automatically. This is documented.

**Reply by RobertAtQuestacon** [reply]
Thanks @rkompass for the prompt reply and you've solved my problem. :)
I've been using the rshell / REPL interface as it matches the way my brain works better. To start the program with REPL, Cntrl-D is pressed, which issues a soft-reset. This does not fix the stuck DMA issue.
I reloaded your unmodified example and it worked first time this morning, but after I stopped it before completion using Cntrl-C and then tried to restart it with Cntrl-D, the transfer count is stuck where I interrupted things.
I have discovered that I get the same behaviour using Thonny by pressing the Stop button before the countdown has completed. I am using a basic Raspberry Pi Pico, so you should be able to repeat the issue.
I tried adding restarts for both sm and dma in all sorts of places without effect.
If I add a sm.put after sm.start, then I see that the transfer count will continue down from where it had been stuck.
If I let the program run to completion, it will then run again each time without the need for a the sm.put().
To use deinit(), as you suggested, I created the dma object, then used deinit, then recreated the object. This worked. Yay! :)
Looking into deinit, I see that it calls abort(), and by adding this instead, the program never gets stuck. I also checked that it uses the same DMA channel each time.
Thanks again. 
I can't claim to entirely understand things. It appears that the dma registers can retain values for sometime after the pico has been switched off. 
Looking at "2.5.5.3. Channel Abort" of the RP2040 documentation, I see that aborting a dma process is not trivial. Possibly, I do not want to always use abort before dma.start(), but I shall do so until it causes me problems. :)

**Reply by RobertAtQuestacon** [reply]
Eight channels of dma to pio to apa102 LEDS working. Here's the code so far for anyone trying to do that same:
```
import rp2
from machine import Pin

@rp2.asm_pio(out_shiftdir=0, autopull=True, pull_thresh=32, sideset_init=rp2.PIO.OUT_LOW, out_init=rp2.PIO.OUT_LOW)
def spi_cpha0():
    out(pins, 1)      .side(0x0)
    nop()             .side(0x1)
    
from rp2 import StateMachine

class SPI:
    #from https://github.com/raspberrypi/pico-micropython-examples/issues/20
    
    #FIXME: Generalize to support enabling the "cpha" and "cpol" parameters.
    def __init__(
        self,
        pio_statemachine_id,
        mosi,
        sck,
        baudrate,
        cpha=False,
        cpol=False,
    ):
        # If this SPI baudrate exceeds the Pico's system clock frequency,
        # raise an exception.
        if baudrate > 125000000:
            raise ValueError(
                f'SPI baudrate {baudrate} > system clock frequency {BAUDRATE_MAX}.')
         
        # Create, classify, and enable a new FSM running this PIO assembly.
        self._sm = StateMachine(
            # These first two parameters *MUST* be passed positionally.
            pio_statemachine_id,
            spi_cpha0,
            # All remaining parameters *MAY* be passed by keyword.
            #FIXME: Why the "4" multiplier magic number here? *sigh*
            freq=4 * baudrate,
            sideset_base=sck,
            out_base=mosi
        )
    #enddef
    
    def active(self,state):
        self._sm.active(state)
    #enddef

    def write(self, buffer_in):        
        for d in buffer_in:
            self._sm.put(d)
        #endfor
    #enddef
            
    def state_machine(self):
      return self._sm
    #enddef

#endclass

import time
from array import array
from rp2_dma import RP2_Dma    # https://github.com/rkompass/RPi2040_mPy/tree/main/rp2_dma

#dma code inspired from https://github.com/orgs/micropython/discussions/11081

if __name__ == "__main__":
    led_strips = []
    dmas = []
    strip1 = array('L',(0x00000000,0xe70000ff,0xe700ff00,0xe7ff0000,0xffffffff))  #red, green, blue
    strip2 = array('L',(0x00000000,0xe700ffff,0xe7ffff00,0xe7ff00ff,0xffffffff))  #yellow, cyan, magenta
    for i in range(8):
      led_strip = SPI(i,Pin(i*2),Pin(i*2+1),2000)
      dma = RP2_Dma(strip1, led_strip.state_machine())  # - a free channel will be assigned
      dma.abort()  #only required if the dma transfer was previously  interrupted.
      dma.start()
      led_strip.active(1)
      #load into list so the garbage collector doesn't gobble them up
      led_strips.append(led_strip)
      dmas.append(dma)
    #endfor
    while True:
      for dma in dmas:
        if dma.busy() == 0:  #has dam transfer finished?
          #led_strips[i].active(0)
          dma.restart()
        #endif
      #endfor
    #endwhile
#endif

```

**Reply by rkompass** [reply]
I had a look at you code: Impressing.
You set up 8 SPI channels for 8 LED strips, which seems crazy :-)
Now you may have 8 data arrays in the end containing the colors of each LED and do computations on these and the DMA will transfer the changes immediately to the strips. Only the DMAs will have to be restarted again and again.
You perhaps want to do this with a periodic interrupt.
If you like it to be more complicated:
As there are only 12 DMA channels this cannot be achieved by cyclic DMA mode (which is an option in my class) as this needs an extra channel to re-trigger each DMA channel. But you may use the class to employ the MULTI_CHAN_TRIGGER register to retrigger all 8 DMAs at once by a single 9-th DMA. There is the `chain_to()` method that is used in cyclic DMA (if you e.g. specify an array of sources for DMA that will be cycled through) that you may use to connect one of the 8 DMA channels to chain to a 9-th channel that you set up so that it writes the bits of the 8 channels into the MULTI_CHAN_TRIGGER register. This will re-trigger all the 8 LED DMA. So this 9-th DMA does not have to be chained to anything else.

**Reply by RobertAtQuestacon** [reply]
We have a light sculpture which has 16 Arduino Dues controlling 192 x 237 addressable LEDs. It works well, most of the time, but I'm hoping to upgrade to Picos which should allow me to add a whole new range of displays.
I was wondering how I was going to trigger the DMA updates. I'd seen mention of chaining and interrupts, but I hadn't got any further than that.  I'll definitely try to implement your suggestion.

**Reply by RobertAtQuestacon** [reply]
I had a go at chaining the dma 0 to dma 9, but I don't think I understand cyclic operation well enough. dma9 retriggers some or all of the other dmas (possibly too soon for some to be ready) but I don't think any of the dmas are cycling their source addresses.
Here is where I got to:
```
import time
from array import array
from rp2_dma import RP2_Dma    # https://github.com/rkompass/RPi2040_mPy/tree/main/rp2_dma

#dma code inspired from https://github.com/orgs/micropython/discussions/11081

@micropython.viper
def chaintrig():
    multi_trig = array('L',(0x000000ff,))  #the first 8 DMAs to retrigger
    dma_trig = RP2_Dma(multi_trig, ptr32(0x50000430),chan=9)  #should only transfer on value and trig restart of others
    return dma_trig
#enddef

def allStart():
  machine.mem32[0x50000430] = 0xff
#enddef  

def setCyclic(chan,cyclic_size,src_inc = 1):
      dma_addr = 0x50000000 + chan * 0x40
      print("Before:",machine.mem32[dma_addr + 16])
      machine.mem32[dma_addr + 16] &= 0xffffc2ff  #remove bit 4,6,7,8,9
      machine.mem32[dma_addr + 16] |= (1 << 10)  |  (cyclic_size << 6) | (src_inc << 4)
      print("After:",machine.mem32[dma_addr + 16])
#enddef

if __name__ == "__main__":
    led_strips = []
    dmas = []
    strips = []
    for i in range(8):
      led_strip = SPI(i,Pin(i*2),Pin(i*2+1),2000)
      strip = array('L',(0x00000000,0xe70000ff,0xe700ff00,0xe7ff0000,0xfff000f0,0xffffffff,0x00000000,0x00000000))  #red, green, blue
      strips.append(strip)
      dma = RP2_Dma(strip, led_strip.state_machine(),chan=i)  # - a free channel will be assigned
      setCyclic(i,3)  #repeat on 8 values
      dma.abort()  #only required if the dma transfer was previously  interrupted.
      #dma.start() - will use allStart function instead
      led_strip.active(1)
      #load into list so the garbage collector doesn't gobble them up
      led_strips.append(led_strip)
      dmas.append(dma)
    #endfor
    dma9 = chaintrig()  #channel 9 retriggers first 8 channels
    setCyclic(9,0,0)  #repeat the same value
    dma9.abort()
    dmas[0].chain_to(dma9)
    
    for i in range(8):
      strip = strips[i]
      strip[2] = 0xe000ffff  | i<<24  #adjust brightness of green
    allStart()
    while True:
      #for dma in dmas[4:]:  #only restart the last 4 channels
      #  if dma.busy() == 0:  #has dma transfer finished?
      #    #led_strips[i].active(0)
      #    dma.restart()
      #  #endif
      #endfor
      led1 = strips[0][1]
      #print("%x"%(led1,))
      #if (led1 & 0xff) == 0xff:
      #  strips[0][1] = 0xe700ff00
      #else:
      #  strips[0][1] = 0xe70000ff
      #endif
    #endwhile
#endif
```
You'll see that I've lengthened the source material to 8 values so that it can cycle on an 8 value boundary.
I don't think I've set the 9th dma to not increment correctly.

If it turns out that this automatic retrigger isn't possible, then setting an IRQ on DMA completion would be fine, but that's a whole nuther rabbit hole. :)

**Reply by rkompass** [reply]
I tried to make a simple demonstration with a PIO state machine that does a single blink and then waits (blockingly) for data.
And found that the re-triggerring with the MULTI_CHAN_TRIGGER register works but thinks are still not that simple.
The code is here. You might connect a LED to pin 3 or change for pin 25 (ordinary Pico, not W):
```py
from time import sleep_ms
from machine import Pin, mem32
from rp2 import PIO, StateMachine, asm_pio
from rp2_dma import RP2_Dma    # https://github.com/rkompass/RPi2040_mPy/tree/main/rp2_dma
from array import array

@rp2.asm_pio(set_init=rp2.PIO.OUT_LOW)
def blink_once():
    wrap_target()
    pull(block)
    set(pins, 1)   [31]
    nop()          [31]
    nop()          [31]
    nop()          [31]
    nop()          [31]
    nop()          [31]
    nop()          [31]
    nop()          [31]
    nop()          [31]
    nop()          [31]
    set(pins, 0)   [31]
    nop()          [31]
    nop()          [31]
    nop()          [31]
    nop()          [31]
    nop()          [31]
    nop()          [31]
    nop()          [31]
    nop()          [31]
    nop()          [31]
    wrap()

# state machine with a single blink program, at 2000Hz, to Pin(3); 
sm = rp2.StateMachine(0, blink_once, freq=2000, set_base=Pin(3))
sm.active(1)  # make it active; each sm.put(x) will lead to a single blink

arr = array('L', (0, 0, 0, 0, 0))
dma = RP2_Dma(arr, sm, chan=8)
dma.start()
print('DMA channel 8 started')
sleep_ms(3000)

mem32[0x50000430] = 1<<8
print('triggered DMA channel 8 again')
sleep_ms(3000)

retrig = array('L', (1<<8,))
dma_retrig = RP2_Dma(retrig, 0x50000430, rw_incr=(0,0), chan=9)
dma.chain_to(dma_retrig)
dma.restart()
print('chained DMA to retrigger-DMA and restarted')

while True:
    print('dma.read_addr():', dma.read_addr())
    sleep_ms(1000)
```
Output:
```
DMA channel 8 started
triggered DMA channel 8 again
chained DMA to retrigger-DMA and restarted
dma.read_addr(): 536953508
dma.read_addr(): 536953520
dma.read_addr(): 536953532
dma.read_addr(): 536953544
dma.read_addr(): 536953556
dma.read_addr(): 536953568
dma.read_addr(): 536953580
dma.read_addr(): 536953592
dma.read_addr(): 536953604
dma.read_addr(): 536953620
. . . .
```
The critical thing is: When the DMA transfer is finished (all the preprogrammed (transfer-count) transfers are done), the count register resets itself to the previously programmed value. But the read and write addresses do not reset. They are at the end of the array, if there was a read or write increment set.
So a simple re-triggering does not work.
It seems to work in the above example, because the re-triggering DMA is set to have read and write increments 0. And because the state machine program does not care what data it is being send to. The printing shows that after the re-triggering of the `dma` the read addresses are past the array `arr`.

So instead of just re-triggering one would need to write the read address through the triggering alias READ_ADDR_TRIG to `dma`.
As this has to happen in your case for all 8 DMA channels the re-triggering DMA has to write an array of the 8 addresses of the led_strips[0] ..led_strips[7] to the 8 addresses of the READ_ADDR_TRIG of the 8 DMA channels.
The latter are not contiguous, therefore this would have to happen through a write to the WRITE_ADDR_TRIG of this re-triggering DMA by a secondary re-triggering DMA that reads from an array of the 8 READ_ADDR_TRIG addresses.
This still is possible. In total you use already 10 of the 12 DMA channels.

I noted that on my Pico W the DMA channel 0 is not free. I assume it is used for SPI (?) communication with the WLAN chip. 

In the end having a timer interrupt that does a `dma.restart()` for all you channels seems appropriate, unless you want to learn by experimentation.

**Reply by RobertAtQuestacon** [reply]
This looks like a good base for my experiments,
BTB I killed my micropython last Friday, so I reinstalled micropython (RPI_PICO-20231005-v1.21.0) and now when I use rp2_dma.py, it complains that it couldn't find a ctypes module. So I changed your import from cytpes to uctypes and it is happy again.
When I ran your retrigger dma test, I got:
```
chained DMA to retrigger-DMA and restarted
dma.read_addr(): 536916944
dma.read_addr(): 536916960
dma.read_addr(): 536916972
dma.read_addr(): 536916984
dma.read_addr(): 536916996
dma.read_addr(): 536917008
```
I'm not surprised that the addresses are different, but for some reason, the first retrigger got an increment of 16 then 12 forever afterwards, whereas your increments were consistently 12.
I can't explain that behaviour, but never mind.
Out of curiosity, I put some read_addr() requests before running the dma, but that froze the read address at 536916912 -- yet more rabbit holes.
I might have one more go at using cyclic addresses then use a timer interrupt as you suggested. :)

P.S. I just reinstalled your original test. When I run the first time, I now get a consistent step of 12, but if I run it a second time, I get no step at all (until I power off the device and run rshell again). Obviously, a soft reset is not stopping something.

**Reply by RobertAtQuestacon** [reply]
I used the first example from your library to test ring address wrapping. I have been wrong in calling it cyclic addressing (I don't know where I got that name from).

```
# Test 1 of RP2_Dma micropython class
#
# In this test we investigate ordinary memory array transfer
#   and measure memory consumption and timing of the dma
#

from array import array
from time import sleep_ms, ticks_us, ticks_diff
import gc

#@micropython.viper
def setRingAddressing(chan,cyclic_size,src_inc = 1):
      dma_addr = 0x50000000 + chan * 0x40
      print("Before:0x%x"% machine.mem32[dma_addr + 16])
      machine.mem32[dma_addr + 16] &= 0xfffff83f  #remove bit 6,7,8,9,10
      #bit 10 is 1 for write address wrap
      machine.mem32[dma_addr + 16] |=  (cyclic_size << 6)  
      print("After:0x%x"% machine.mem32[dma_addr + 16])
#enddef


# import machine
# machine.freq(250_000_000)   # of course works twice as fast

gc.collect(); mem0 = gc.mem_free()
print('before "import RP2_Dma":', mem0)
from rp2_dma import RP2_Dma

gc.collect(); mem1 = gc.mem_free()
print('after "import RP2_Dma":', mem1, 'consumed:', mem0-mem1, 'bytes')

# barr = bytearray((0,)*400)
sarr = array('i', (i for i in range(10)))
print("sarr = array('i', (i for i in range(-10, 9_990)))")

darr = array('L', (0 for _ in range(10)))   
print("darr = array('L', (0 for _ in range(12_000)))")

gc.collect(); mem2 = gc.mem_free()
print('after seeting up arrays:', mem2, 'consumed:', mem1-mem2, 'bytes')

# ----------------------------------------------
dma = RP2_Dma(sarr,  darr, chan=8)
# ----------------------------------------------
# dsize is determined automatically from array type as 2 (=32 bit)
# alternatively you may specify dsize explicitly and try out:
# dma = RP2_Dma(sarr,  darr, dsize=0)   # byte transfer, or
# dma = RP2_Dma(sarr,  darr, dsize=1)   # 16-bit transfer
# if dsize is too small we might only transfer half or quarter of the data

src_start = dma.read_addr()
print('Src start addr:', src_start)

setRingAddressing(8,4)  #2 repeats 1 src, 3 repeats 2 srcs, 4 = 4, 5 = 8 values

print()

gc.collect(); mem3 = gc.mem_free()
print('dma = RP2_Dma(sarr,  darr):', mem3, 'consumed:', mem2-mem3, 'bytes')

t_sta = ticks_us()
dma.start()
while dma.busy():
    pass
t_fin = ticks_us()
t_dif = ticks_diff(t_fin, t_sta)
print('dma transfer took', t_dif, 'us for 10000 elements:', t_dif/10, ' ns per single transfer')
print()
print('Src fini addr:', dma.read_addr(), " Delta:", dma.read_addr() - src_start)

print('darr:', darr)
```
This works for ring values 2,3 & 4 which give 1,2 and 4 step wraps. This doesn't match the RP2040 documentation. For values greater than 4, it appears to attempt to do the right thing, but after 4 transfers, the data is corrupted.
The output for the RING_SIZE = 4 is shown below.

```
MPY: soft reboot
before "import RP2_Dma": 219328
after "import RP2_Dma": 209744 consumed: 9584 bytes
sarr = array('i', (i for i in range(-10, 9_990)))
darr = array('L', (0 for _ in range(12_000)))
after seeting up arrays: 209440 consumed: 304 bytes
Using dma chan  8
Integer DMA settings: array('i', [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]) 0 None None None
Integer DMA settings: array('L', [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]) 0 None None None
we configure this single (perhaps slave) dma
Src inc: 1  Dest inc: 1
tr_cnt 10
Src start addr: 536911856
Before:0x3fc03b
After:0x3fc13b

dma = RP2_Dma(sarr,  darr): 209488 consumed: -48 bytes
dma transfer took 185 us for 10000 elements: 18.5  ns per single transfer

Src fini addr: 536911864  Delta: 8
darr: array('L', [0, 1, 2, 3, 0, 1, 2, 3, 0, 1])
```

It would have been good if it worked. I could have sent 256 LED values (I only needed 237) repeatedly using the retriggering (assuming the two things didn't interfere).
I may follow this up sometime, but at this stage I'll move onto timer interrupts for retriggering my LED dma.

**Comment by HughMaingauche** [comment]
Probably too late but I wrote 2 years ago a simple VGA driver for micropython on the PICO using PIO and DMA. Synchronization between PIO and DMA is achived using PIO Irqs.

This might provide you with interesting examples. Repo on Github [here](https://github.com/HughMaingauche/PICO-VGA-Micropython)

Link [here ](https://forum.micropython.org/viewtopic.php?f=15&t=12094&p=66142&hilit=vga#p66142)in the old micropython forum

Hugh
